File: python/py_sparker/converters.py
from objects import Profile, MatchingEntities, KeyValue, BlockClean, BlockDirty, BlockWithComparisonSize, ProfileBlocks
from blockers import TokenBlocking

class Converters(object):
	"""
		Contains converters to convert RDDs to different object types
	"""
	@staticmethod
	def blockIDProfileIDFromBlock(block):
		"""
		Given a block (clean or dirty), produces a list of (p_i, BlockWithComparisonSize)
		where p_i is a profile indexed in the block.
		
		Parameters
		----------
		block : BlockClean|BlockDirty
			A block (clean or dirty)
		"""
		blockWithComparisonSize = BlockWithComparisonSize(block.blockID, block.getComparisonSize())
		return map(lambda p: (p, blockWithComparisonSize), block.getAllProfiles())

	@staticmethod
	def blocksToProfileBlocks(blocks):
		"""
		Given an RDD of blocks (clean or dirty), produces an RDD of ProfileBlocks.
		
		Parameters
		----------
		block : RDD[BlockClean|BlockDirty]
			An RDD of blocks (clean or dirty)
		"""
		profilesPerBlocks = blocks.flatMap(Converters.blockIDProfileIDFromBlock).groupByKey()
		return profilesPerBlocks.map(lambda b : ProfileBlocks(b[0], set(b[1])))

	@staticmethod	
	def profilesBlockToBlocks(profilesBlocks, separatorIDs = []):
		"""
		Given an RDD of ProfileBlocks, produces an RDD of blocks (clean or dirty).
		
		Parameters
		----------
		profilesBlocks : RDD[ProfileBlocks]
			An RDD of profile blocks
		separatorIDs : [int]
			Id of separators (if the ER is clean) that identifies how to split profiles across the different sources
		"""
		def getBlock(data):
			blockID = data[0]
			profilesIDs = set(data[1])
			if len(separatorIDs) == 0:
				return BlockDirty(blockID, [profilesIDs])
			else:
				return BlockClean(blockID, TokenBlocking.separateProfiles(profilesIDs, separatorIDs))
				
		blockIDProfileID = profilesBlocks.flatMap(lambda pb: [(b.blockID, pb.profileID) for b in pb.blocks])
		blocks = blockIDProfileID.groupByKey().map(getBlock)
		return blocks.filter(lambda b: b.getComparisonSize() > 0)

File: scala/sparker/src/main/scala-2.11/SparkER/Utilities/Converters.scala
package SparkER.Utilities

import SparkER.BlockBuildingMethods.TokenBlocking
import org.apache.spark.rdd.RDD
import SparkER.DataStructures._
import org.apache.spark.partial.PartialResult

/**
  * Contains differents methods
  * to convert to differents DataStructures types.
  *
  * @author Luca Gagliardelli
  * @since 2016/12/08
  */
object Converters {

  /**
    * Given a RDD of blocks return a RDD of profiles block
    *
    * @param blocks RDD of blocks
    * @return RDD of profile blocks
    **/
  def blocksToProfileBlocks(blocks: RDD[BlockAbstract]): RDD[ProfileBlocks] = {
    val profilesPerBlocks = blocks.flatMap(blockIDProfileIDFromBlock).groupByKey()
    profilesPerBlocks map (x => ProfileBlocks(x._1, x._2.toSet))
  }

  /**
    * Given a block return a list that contains for each profile in the block (profileID, (blockID, number of comparison in the block))
    *
    * @param block block
    * @return a list that contains for each profile in the block (profileID, (blockID, number of comparison in the block))
    **/
  def blockIDProfileIDFromBlock(block: BlockAbstract): Iterable[(Long, BlockWithComparisonSize)] = {
    val blockWithComparisonSize = BlockWithComparisonSize(block.blockID, block.getComparisonSize())
    block.getAllProfiles.map((_, blockWithComparisonSize))
  }


  /**
    * Given a RDD of profiles block return a RDD of blocks
    *
    * @param profilesBlocks RDD of profileBlock
    * @param separatorIDs max ID of the first dataset (if it is clean-clean context), default -1
    * @return RDD of blocks
    **/
  def profilesBlockToBlocks(profilesBlocks: RDD[ProfileBlocks], separatorIDs: Array[Long] = Array.emptyLongArray): RDD[BlockAbstract] = {

    val blockIDProfileID = profilesBlocks flatMap {
      profileWithBlocks =>
        val profileID = profileWithBlocks.profileID
        profileWithBlocks.blocks map {
          BlockWithSize =>
            (BlockWithSize.blockID, profileID)
        }
    }

    val blocks = blockIDProfileID.groupByKey().map {
      block =>
        val blockID = block._1
        val profilesID = block._2.toSet

        if (separatorIDs.isEmpty) {
          BlockDirty(blockID, Array(profilesID))
        }
        else {
          BlockClean(blockID, TokenBlocking.separateProfiles(profilesID, separatorIDs))
        }
    }

    blocks.filter(_.getComparisonSize() > 0).map(x => x)

  }
}
